netebpfext/user/fwp_um.h

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+#pragma once
+#include <mutex>
+#include <unordered_map>
+
+typedef class _fwp_engine
+{
+  public:
+    _fwp_engine() = default;
+
+    uint32_t
+    add_fwpm_callout(const FWPM_CALLOUT0* callout)
+    {
+        std::unique_lock l(lock);
+        uint32_t id = next_id++;
+        fwpm_callouts.insert({id, *callout});
+        return id;
+    }
+
+    bool
+    remove_fwpm_callout(size_t id)
+    {
+        std::unique_lock l(lock);
+        return fwpm_callouts.erase(id) == 1;
+    }
+
+    uint32_t
+    add_fwps_callout(const FWPS_CALLOUT3* callout)
+    {
+        std::unique_lock l(lock);
+        uint32_t id = next_id++;
+        fwps_callouts.insert({id, *callout});
+        return id;
+    }
+
+    bool
+    remove_fwps_callout(size_t id)
+    {
+        std::unique_lock l(lock);
+        return fwps_callouts.erase(id) == 1;
+    }
+
+    uint32_t
+    add_fwpm_filter(const FWPM_FILTER0* filter)
+    {
+        std::unique_lock l(lock);
+        uint32_t id = next_id++;
+        fwpm_filters.insert({id, *filter});
+        return id;
+    }
+
+    bool
+    remove_fwpm_filter(size_t id)
+    {
+        std::unique_lock l(lock);
+        return fwpm_filters.erase(id) == 1;
+    }
+
+    uint32_t
+    add_fwpm_sub_layer(const FWPM_SUBLAYER0* sub_layer)
+    {
+        std::unique_lock l(lock);
+        uint32_t id = next_id++;
+        fwpm_sub_layers.insert({id, *sub_layer});
+        return id;
+    }
+
+    bool
+    remove_fwpm_sub_layer(size_t id)
+    {
+        std::unique_lock l(lock);
+        return fwpm_sub_layers.erase(id) == 1;
+    }
+
+    FWP_ACTION_TYPE
+    classify_test_packet(_In_ const GUID* layer_guid, NET_IFINDEX if_index);
+
+    static _fwp_engine*
+    get()
+    {
+        if (!_engine)
+            _engine = std::make_unique<_fwp_engine>();
+        return _engine.get();
+    }
+
+  private:
+    _Ret_maybenull_ const FWPM_FILTER*
+    get_fwpm_filter_with_context()
```

Comment:
Why return arbitrary filter here? Probably depending on what we are testing, a specific filter should be returned as the hook providers make use of the filter context, isnt it?

Reply:
Agreed. What is it supposed to match on here? Is it layer_guid against FWPM_FILTER0::layerKey?

Reply:
https://github.com/microsoft/ebpf-for-windows/issues/1474